tools/codegen/api/lazy.py

```diff
@@ -89,45 +88,70 @@ def isWrappedScalarType(typ: Type) -> bool:
         return typ.name == BaseTy.Scalar
     elif isinstance(typ, (OptionalType, ListType)):
         return isWrappedScalarType(typ.elem)
-    else:
-        return False
+    return False
+
+def isGeneratorType(typ: Type) -> bool:
+    if isinstance(typ, BaseType):
+        return typ.name == BaseTy.Generator
+    elif isinstance(typ, (OptionalType)):
+        return isGeneratorType(typ.elem)
+    return False
+
+class LazyArgument:
+    name: str
+    orig_type: Type
+    lazy_type: CType
+    is_wrapped_scalar: bool
+    is_generator: bool
+
+    # true if this argument is or contains a lazy IR value
+    is_lazy_value: bool
+
+    def __init__(self, arg: Argument):
+        self.name = arg.name
+        self.orig_type = arg.type
+        self.is_generator = isGeneratorType(arg.type)
+        if self.is_generator:
+            self.lazy_type = None
+        else:
+            self.lazy_type = process_ir_type(arg.type)
+        self.is_wrapped_scalar = isWrappedScalarType(arg.type)
 
+        self.is_lazy_value = isValueType(self.lazy_type)
 
 # Inspired by a FunctionSchema object, a LazyIrSchema holds the schema of a Lazy IR node.
 # Unlike a FunctionSchema, it has no round-trippable string form (relating to the YAML),
 # but carries type information from a native FunctionSchema modified for use with IR nodes,
 # and preserving original argument names.
-
-
 class LazyIrSchema:
     # The name of the operator this function schema describes.
     name: 'OperatorName'
 
-    positional_arg_types: Tuple[NamedCType, ...]
-    keyword_arg_types: Tuple[NamedCType, ...]
+    positional_args: Tuple[LazyArgument, ...]
+    keyword_args: Tuple[LazyArgument, ...]
 
     # TODO: Need to handle collisions with argument names at some point
     returns: Tuple['Return', ...]
 
-    wrapped_scalar_names: List[str]
+    # if this schema has a Generator arg, list its orig ctype/name but don't
+    # build a LazyArgument since lazy IR doesn't support it
```

Comment:
Is there a downside to just keeping this in `positional_args/keyword_args`, and not bothering to special case generator in the data model?

If you need to emit special fallback code for ops that have a generate argument, you could do something like:
```
if any(isinstance(a, OptionalType) and a.elem is BaseTy.Generator for a in itertools.chain(positional_args, keyword_args)
```

(And if you end up needing that in multiple places you could always make it a method on the `LazyIrSchema` class)

Reply:
The PR comment is good, but it might also be worth commenting somewhere in the code - the reason the LTC codegen needs to special-case the `generator` type is because that type isn't represented in LTC's IR (if I understood that correctly)

Reply:
I need to 
 1 - emit special fallback code (a) add 'generator.has_value()...' to fallback conditional, (b) include 'generator' in args to call_fallback_fn
 2 - include generator in calls to compute_shape/meta
 3 - NOT include generator in LazyIR def/decl/call/lowering

So for 1, 2 I could simplify things
but for 3 i have to really have the filtering ability somehow